#43 write outputs for displayQA, add rho1, and make it work with the latest reductions

HironaoMiyatake opened the pull request:
Sorry that I ignored this issue for a while. There will be several things to be improved to have Stile work with displayQA more nicely (now I can generate decent displayQA with this pull request, though), but I would like to make a pull request now since I already made a bunch of changes.

- I added a class to calculate rho1
- Data points used for whisker plots and scatter plots are written down into a file.
- I needed to tweak several things to make Stile work with the latest pipeline products. One thing I should notice here is that I had to ignore a patch which does not exist as a catalog but does exist in `dataRefList` (l.728 in base_tasks.py). This should be removed when the HSC pipeline is fixed.

msimet commented:
I just noticed a kind of silly thing: I think I was calling these Rowe statistics (after @barnabytprowe) not rho statistics.  @rmandelb, I know we've discussed this before--did we ever come to a conclusion about which thing people meant?

Otherwise, cool!  I'll try to look this over soon, thanks.

rmandelb commented:
Well, they are defined in [a single-author paper by Barney](http://adsabs.harvard.edu/abs/2010MNRAS.404..350R).  And for example [this paper](http://physics.ohio-state.edu/~msgill/astro/A611_draft.pdf) describes them as "Rowe" statistics.  But I've seen others who wrote rho.  So basically, I have no clue.  Sorry, I'm not helpful at all.

barnabytprowe commented:
Call them rho statistics :)  It's a pun anyway, so should always be preferred over something more pious, and I'm happier with a greek letter getting assigned given that all the paper discussed was something so simple as a auto- and cross correlations constructed from modelling residuals!

barnabytprowe commented:
(Sorry, wouldn't have seen this but for the ping ;)

msimet commented:
Well, I pinged you for a reason. :) I was actually just coming over to say I'd found an old email where you said you didn't like that name so we should stick with rho.  So I think that's the executive decision, then.

msimet commented on stile/sys_tests.py:
```diff
@@ -859,12 +894,20 @@ class WhiskerPlotStarSysTest(WhiskerPlotSysTest):
                                 size_label = r'$\sigma$ [pixel]',
                                 xlim = xlim, ylim = ylim, equal_axis = True)
 
+    def getData(self):
```
You could make this function a function of `WhiskerPlotSysTest` and then it wouldn't need to be repeated in all the subclasses

HironaoMiyatake replied:
Right, I'll move that this function to `WhiskerPlotSysTest`.

msimet commented on stile/sys_tests.py:
```diff
@@ -902,6 +963,106 @@ class ScatterPlotSysTest(SysTest):
+                x, y, yerr = self.getStatisticsPerCCD(array['CCD'], array[x_field],
+                                                      array[y_field], yerr = array[yerr_field],
+                                                      stat = per_ccd_stat)
+                self.data = stile_utils.FormatArray(numpy.rec.fromarrays([list(set(array['CCD'])), x,
```
You don't need to run this through `FormatArray`, I think, since it's doing almost the same thing as `numpy.rec.fromarrays`--unless you tried it without and it didn't work?

HironaoMiyatake replied:
I found that it works without running `FormatArray`. I'll just use `numpy.rec.fromarrays`.

msimet commented on stile/sys_tests.py:
```diff
@@ -1157,17 +1338,15 @@ class ScatterPlotStarVsPSFG1SysTest(ScatterPlotSysTest):
-                                xlabel=r'$g^{\rm PSF}_1$', ylabel=r'$g^{\rm star}_1$',
-                                color=color, lim=lim, equal_axis=False,
-                                linear_regression=True, reference_line='one-to-one')
+        self.per_ccd_stat = None if per_ccd_stat == 'None' else per_ccd_stat
```
Is there a reason the default is `'None'` rather than just `None`?

HironaoMiyatake replied:
This is because when we pass `None` through the command option in HSC pipeline, it's treated as string. However, I think this line should be put into `hsc/sys_test_adapters.py` as it's related to the pipeline. I've done this change.

msimet commented on stile/sys_tests.py:
```diff
@@ -1157,17 +1338,15 @@ class ScatterPlotStarVsPSFG1SysTest(ScatterPlotSysTest):
-                                color=color, lim=lim, equal_axis=False,
-                                linear_regression=True, reference_line='one-to-one')
+        self.per_ccd_stat = None if per_ccd_stat == 'None' else per_ccd_stat
+        return super(ScatterPlotStarVsPSFG1SysTest, self).__call__(array, 'psf_g1', 'g1', 'g1_err',
```
This reorganization is nifty!

HironaoMiyatake replied:
:blush:

msimet commented on stile/hsc/base_tasks.py:
```diff
@@ -832,17 +857,35 @@ def _makeArgumentParser(cls):
         dtype = str,
         default = "deep",
     )
+    # Generate a list of flag columns to be used in the .removeFlaggedObjects() method
```
Is there a reason to add a copy of this here?  It should just inherit directly from `CCDSingleEpochStileConfig`.

HironaoMiyatake replied:
Now I added 'flags.pixel.clipped.any' here. This flag does not exist in visit-level, so I think I need to copy this.

msimet commented:
Hey @HironaoMiyatake -- this looks good; I've left a couple of comments but they shouldn't affect functionality.  I also made a commit updating a bit of the wording in the documentation (4a4743b).

I merged master back into this branch, too.  The only change you'll need to make, I think, is that if we really do need to keep the individual `flags` arguments in the different `Config` objects, they'll need to be changed to `flags_keep_false`.

HironaoMiyatake commented:
@msimet , thank you for reviewing the code! I went through your comments.
I made two additional changes;
- added `flags.pixel.clipped.any` to tract level configs
- added CCD type in `VisitSingleEpochStileConfig`
.

Please let me know if you are happy with current version. Then I'll merge this into master.

msimet commented:
Looks good to me, feel free to merge!

HironaoMiyatake commented:
I just merged. I'll look into PR#5 then!
